File: js_project/react-app/src/lib/chainOfResponsibility.js
/**
 * Паттерн Chain of Responsibility для обработки начислений.
 */

import { Charge, Service, PaymentNotice } from './models.js';

/**
 * Абстрактный обработчик начислений.
 */
export class ChargeHandler {
    constructor() {
        /** @type {ChargeHandler|null} */
        this._nextHandler = null;
    }

    /**
     * Устанавливает следующий обработчик в цепочке.
     * @param {ChargeHandler} handler - Следующий обработчик
     * @returns {ChargeHandler} Следующий обработчик (для цепочки вызовов)
     */
    setNext(handler) {
        this._nextHandler = handler;
        return handler;
    }

    /**
     * Обрабатывает начисление.
     * @param {Charge} charge - Начисление
     * @param {Service} service - Услуга
     * @returns {number|null} Результат обработки или null
     */
    handle(charge, service) {
        throw new Error("Метод handle должен быть переопределен");
    }

    /**
     * Передает обработку следующему обработчику в цепочке.
     * @param {Charge} charge - Начисление
     * @param {Service} service - Услуга
     * @returns {number|null} Результат обработки или null
     */
    _processNext(charge, service) {
        if (this._nextHandler) {
            return this._nextHandler.handle(charge, service);
        }
        return null;
    }
}

/**
 * Обработчик валидации данных.
 */
export class ValidationChargeHandler extends ChargeHandler {
    /**
     * Валидирует данные начисления.
     * @param {Charge} charge - Начисление
     * @param {Service} service - Услуга
     * @returns {number|null} Результат обработки
     */
    handle(charge, service) {
        if (charge.quantity < 0) {
            throw new Error(`Отрицательное количество для начисления #${charge.chargeCode}`);
        }
        if (service.tariff < 0) {
            throw new Error(`Отрицательный тариф для услуги #${service.serviceCode}`);
        }
        return this._processNext(charge, service);
    }
}

/**
 * Обработчик применения скидки.
 */
export class DiscountChargeHandler extends ChargeHandler {
    /**
     * @param {number} discountThreshold - Порог для применения скидки
     * @param {number} discountPercent - Процент скидки
     */
    constructor(discountThreshold, discountPercent) {
        super();
        this.discountThreshold = discountThreshold;
        this.discountPercent = discountPercent;
    }

    /**
     * Применяет скидку, если сумма превышает порог.
     * @param {Charge} charge - Начисление
     * @param {Service} service - Услуга
     * @returns {number|null} Результат обработки
     */
    handle(charge, service) {
        const baseCost = service.calculateCost(charge.quantity);
        if (baseCost >= this.discountThreshold) {
            const discount = baseCost * (this.discountPercent / 100);
            return baseCost - discount;
        }
        return this._processNext(charge, service);
    }
}

/**
 * Обработчик стандартного расчета.
 */
export class StandardChargeHandler extends ChargeHandler {
    /**
     * Выполняет стандартный расчет стоимости.
     * @param {Charge} charge - Начисление
     * @param {Service} service - Услуга
     * @returns {number} Стоимость
     */
    handle(charge, service) {
        return service.calculateCost(charge.quantity);
    }
}

/**
 * Процессор начислений.
 */
export class ChargeProcessor {
    constructor() {
        const validation = new ValidationChargeHandler();
        const discount = new DiscountChargeHandler(5000.0, 5.0);
        const standard = new StandardChargeHandler();
        
        validation.setNext(discount).setNext(standard);
        this._handler = validation;
    }

    /**
     * Обрабатывает одно начисление.
     * @param {Charge} charge - Начисление
     * @param {Service} service - Услуга
     * @returns {number} Стоимость
     */
    processCharge(charge, service) {
        const result = this._handler.handle(charge, service);
        if (result === null) {
            throw new Error(`Не удалось обработать начисление #${charge.chargeCode}`);
        }
        return result;
    }

    /**
     * Обрабатывает извещение на оплату.
     * @param {PaymentNotice} notice - Извещение на оплату
     * @returns {PaymentNotice} Обработанное извещение
     */
    processNotice(notice) {
        let totalAmount = 0.0;
        for (const [charge, service] of notice.charges) {
            totalAmount += this.processCharge(charge, service);
        }
        notice.totalAmount = totalAmount;
        return notice;
    }
}

File: chain_of_responsibility.py
"""Паттерн Chain of Responsibility для обработки начислений."""

from abc import ABC, abstractmethod
from typing import Optional
from models import Charge, Service, PaymentNotice


class ChargeHandler(ABC):
    """Абстрактный обработчик начислений."""

    def __init__(self):
        self._next_handler: Optional['ChargeHandler'] = None

    def set_next(self, handler: 'ChargeHandler') -> 'ChargeHandler':
        self._next_handler = handler
        return handler

    @abstractmethod
    def handle(self, charge: Charge, service: Service) -> Optional[float]:
        pass

    def _process_next(self, charge: Charge, service: Service) -> Optional[float]:
        if self._next_handler:
            return self._next_handler.handle(charge, service)
        return None


class ValidationChargeHandler(ChargeHandler):
    """Валидация данных."""

    def handle(self, charge: Charge, service: Service) -> Optional[float]:
        if charge.quantity < 0:
            raise ValueError(f"Отрицательное количество для начисления #{charge.charge_code}")
        if service.tariff < 0:
            raise ValueError(f"Отрицательный тариф для услуги #{service.service_code}")
        return self._process_next(charge, service)


class DiscountChargeHandler(ChargeHandler):
    """Применение скидки."""

    def __init__(self, discount_threshold: float, discount_percent: float):
        super().__init__()
        self.discount_threshold = discount_threshold
        self.discount_percent = discount_percent

    def handle(self, charge: Charge, service: Service) -> Optional[float]:
        base_cost = service.calculate_cost(charge.quantity)
        if base_cost >= self.discount_threshold:
            discount = base_cost * (self.discount_percent / 100)
            return base_cost - discount
        return self._process_next(charge, service)


class StandardChargeHandler(ChargeHandler):
    """Стандартный расчет."""

    def handle(self, charge: Charge, service: Service) -> Optional[float]:
        return service.calculate_cost(charge.quantity)


class ChargeProcessor:
    """Процессор начислений."""

    def __init__(self):
        validation = ValidationChargeHandler()
        discount = DiscountChargeHandler(discount_threshold=5000.0, discount_percent=5.0)
        standard = StandardChargeHandler()
        validation.set_next(discount).set_next(standard)
        self._handler = validation

    def process_charge(self, charge: Charge, service: Service) -> float:
        result = self._handler.handle(charge, service)
        if result is None:
            raise ValueError(f"Не удалось обработать начисление #{charge.charge_code}")
        return result

    def process_notice(self, notice: PaymentNotice) -> PaymentNotice:
        total_amount = 0.0
        for charge, service in notice.charges:
            total_amount += self.process_charge(charge, service)
        notice.total_amount = total_amount
        return notice

